### ISSUE TITLE
Backup fails with invalid type for socket .s.PGSQL.5432

### ISSUE THREAD

**Issue by CluelessTechnologist** [pgbackrest/pgbackrest#803]
1. pgBackRest version: pgBackRest 2.16 

2. PostgreSQL version: 9.6.10

3. Operating system/version:  Ubuntu 18.04.1 LTS

4. Did you install pgBackRest from source or from a package? 
Package from http://apt.postgresql.org/pub/repos/apt/ 

5. Please attach the following as applicable:
[pgclu9-backup.log](https://github.com/pgbackrest/pgbackrest/files/3497199/pgclu9-backup.log)

**pgbackrest.conf on client:** 

```
[global]
repo1-host=server-0523.example.com
repo1-host-user=postgres
log-level-file=detail
repo1-path=/backup/backup_pgbackrest/
repo1-retention-full=1

[pgclu9]
pg1-path=/data/patroni_postgres9
pg1-socket-path=/data/patroni_postgres9
```

**pgbackrest.conf on server:**

```
[global]
repo1-path=/backup/backup_pgbackrest/
repo1-retention-full=2
start-fast=y

[pgclu9]
pg1-host=server-0233.example.com
pg1-path=/data/patroni_postgres9
pg1-socket-path=/data/patroni_postgres9
pg1-host-user=postgres

pg2-host=server-0234.example.com
pg2-path=/data/patroni_postgres9
pg2-socket-path=/data/patroni_postgres9
pg2-host-user=postgres

pg3-host=server-0235.example.com
pg3-path=/data/patroni_postgres9
pg3-socket-path=/data/patroni_postgres9
pg3-host-user=postgres
```

**postgresql.conf patroni config:** 

```
# Do not edit this file manually!
# It will be overwritten by Patroni!
include 'postgresql.base.conf'

archive_command = 'pgbackrest --stanza=pgclu9 archive-push %p'
archive_mode = 'True'
cluster_name = 'postgres-cluster-prod9'
hot_standby = 'on'
listen_addresses = '10.6.13.67'
max_connections = '450'
max_locks_per_transaction = '256'
max_prepared_transactions = '0'
max_replication_slots = '10'
max_wal_senders = '10'
max_worker_processes = '8'
port = '5432'
track_commit_timestamp = 'off'
unix_socket_directories = '.'
wal_keep_segments = '8'
wal_level = 'replica'
wal_log_hints = 'on'
hba_file = '/data/patroni_postgres9/pg_hba.conf'
ident_file = '/data/patroni_postgres9/pg_ident.conf'

```

7. Describe the issue: I have a three node patroni postgrescluster. I have the same pgbackrest setup running with pgbackrest 2.08 without issues backing up.

This is the error message: 

```
-------------------PROCESS START-------------------
2019-08-12 14:53:20.782 P00   INFO: backup command begin 2.16: --pg1-host=server-0233.example.com --pg2-host=server-0234.example.com --pg3-host=server-0235.example.com --pg1-host-user=postgres --pg2-host-user=postgres --pg3-host-user=postgres --pg1-path=/data/patroni_postgres9 --pg2-path=/data/patroni_postgres9 --pg3-path=/data/patroni_postgres9 --pg1-socket-path=/data/patroni_postgres9 --pg2-socket-path=/data/patroni_postgres9 --pg3-socket-path=/data/patroni_postgres9 --repo1-path=/backup/backup_pgbackrest --repo1-retention-full=2 --stanza=pgclu9 --start-fast
2019-08-12 14:53:24.420 P00   WARN: no prior backup exists, incr backup has been changed to full
2019-08-12 14:53:24.825 P00   INFO: execute non-exclusive pg_start_backup() with label "pgBackRest backup started at 2019-08-12 14:53:20": backup begins after the requested immediate checkpoint completes
2019-08-12 14:53:25.027 P00   INFO: backup start archive = 000000170000001A00000032, lsn = 1A/32000028
2019-08-12 14:53:25.276 P00  ERROR: [098]: raised from remote process on 'server-0235.example.com': invalid type for '/data/patroni_postgres9/.s.PGSQL.5432' at /loader/0x559e26d7afe8/pgBackRest/Storage/Storage.pm line 403.
                                         at /loader/0x559e26d7afe8/pgBackRest/Main.pm line 12.
                                        	pgBackRest::Main::__ANON__("PGBRCLIB:98:storage/posix/storage.c:146:invalid type for '/da"...) called at /loader/0x559e26d7afe8/pgBackRest/Storage/Storage.pm line 403
                                        	pgBackRest::Storage::Storage::manifest(undef, "/data/patroni_postgres9", HASH(0x559e27a1a9f0)) called at /loader/0x559e26d7afe8/pgBackRest/Protocol/Remote/Minion.pm line 115
                                        	pgBackRest::Protocol::Remote::Minion::__ANON__(ARRAY(0x559e27a0a900)) called at /loader/0x559e26d7afe8/pgBackRest/Protocol/Base/Minion.pm line 192
                                        	eval {...} called at /loader/0x559e26d7afe8/pgBackRest/Protocol/Base/Minion.pm line 215
                                        	eval {...} called at /loader/0x559e26d7afe8/pgBackRest/Protocol/Base/Minion.pm line 224
                                        	pgBackRest::Protocol::Base::Minion::process(pgBackRest::Protocol::Remote::Minion=HASH(0x559e279d74b0), "/tmp/pgbackrest", "backup", "pgclu9demun1", 0) called at /loader/0x559e26d7afe8/pgBackRest/Main.pm line 106
                                        	eval {...} called at /loader/0x559e26d7afe8/pgBackRest/Main.pm line 238
                                        	pgBackRest::Main::main("remote") called at (eval 276) line 1
                                        	eval '($iResult, $bErrorC, $strMessage) = pgBackRest::Main::main(\'remote\')' called at (eval 276) line 0
                                        
2019-08-12 14:53:25.281 P00   INFO: backup command end: aborted with exception [098]
```

**Comment by dwsteele** [comment]
Locating the socket in PGDATA (`unix_socket_directories = '.'`) is not a good idea.  I'm pretty sure Patroni does not require this (unless something has changed recently).

If you must, you can use `--exclude` (https://pgbackrest.org/configuration.html#section-backup/option-exclude) to exclude this file from the backup but we don't recommend it.

**Comment by bradnicholson** [comment]
> Locating the socket in PGDATA (unix_socket_directories = '.') is not a good idea. I'm pretty sure Patroni does not require this (unless something has changed recently).

I can confirm there is no such requirement from Patroni.  The default `/tmp` location works just fine

**Comment by CluelessTechnologist** [comment]
> > Locating the socket in PGDATA (unix_socket_directories = '.') is not a good idea. I'm pretty sure Patroni does not require this (unless something has changed recently).
> 
> I can confirm there is no such requirement from Patroni. The default `/tmp` location works just fine

@bradnicholson can you share your patroni config? I have not specifced unix_socket_directories in it. Should I add it under parameters and change it to **unix_socket_directories = '/tmp'** ?

This is my current config:
```
patronictl -c /etc/patroni9.yml show-config
loop_wait: 10
maximum_lag_on_failover: 1048576
postgresql:
  parameters:
    archive_command: pgbackrest --stanza=pgclu9 archive-push %p
    archive_mode: true
    max_connections: 450
    max_locks_per_transaction: 256
  use_pg_rewind: true
retry_timeout: 10
ttl: 30
```

**Comment by jpuris** [comment]
Run in exact same problem. Updated patroni.yaml to have a different socket dir
OLD
```
    parameters:
        unix_socket_directories: '., /var/run/postgresql/'
```
NEW
```
    parameters:
        unix_socket_directories: '/tmp/, /var/run/postgresql/'
```
Then restart patroni.

This fixed it.

**Comment by jpuris** [comment]
Forgot to mention.

We are going to use `--exclude` temporarily, until we have good enough reason to restart production cluster primary.

Also when pgbackrest fails like this, it will not release the backup, so you will need to go to clusters primary and issue `select select pg_stop_backup();`. Perhaps would be appropriate to open another issue on this.

**Comment by CluelessTechnologist** [comment]
@jpuris can you share your backup line?
I cannot get it to exclude the socket file.

`su postgres -c "pgbackrest --stanza=pgclu9 --exclude=.s.PGSQL.5432 backup"`

**Comment by jpuris** [comment]
> @jpuris can you share your backup line?
> I cannot get it to exclude the socket file.
> 
> `su postgres -c "pgbackrest --stanza=pgclu9 --exclude=.s.PGSQL.5432 backup"`

I have just tried it and it can not be excluded. Seems the exclusion works only for "valid" files.

If `.s.PGSQL.5432` is a socket file, `--exclude=.s.PGSQL.5432` does not work
If `.s.PGSQL.5432` is a regular file, `--exclude=.s.PGSQL.5432` works
Regular file was created with `touch $PGDATA/.s.PGSQL.5432`

You could remove the socket files, but that will break the connectivity to db via socket file.

**Comment by bradnicholson** [comment]
Correction - actually do have it set to `unix_socket_directories: '/tmp'`

**Comment by dwsteele** [comment]
> I cannot get it to exclude the socket file.

Hmm, yeah, the error for the invalid file type comes before the exclusion.  We'll make this better when we migrate to C, but for now the solution is to keep the socket files out of $PGDATA, which is a good idea in any case.

> Also when pgbackrest fails like this, it will not release the backup

@jpuris This is on purpose, since `pg_stop_backup()` can block indefinitely when there are issues and we feel it is more important to exit with the original error as soon as possible.  On PG <= 9.5 you can use `stop-auto` to automatically stop the old backup when starting a new one.  PG >= 9.6 automatically stops the backup when pgBackRest terminates the connection to PG so this is not an issue.

**Comment by dwsteele** [comment]
Looping in @rachid-casa from #810.

These issues are a change in behavior from 2.15 and are caused by the C code being more correct about what is automatically excluded.  The Perl code should not have been automatically excluding socket files but a mistake in a regexp meant that all files beginning with `.` were being excluded, instead of just `.`.

For 2.17 we are planning to make backup WARN when special files (e.g. socket, pipe) are found and automatically exclude them (as `pg_basebackup` does`).  The WARN can be suppressed using `--exclude`.

The `.s.PGSQL.XXXX.lock` file will be included in the backup unless it is excluded.

Does that sound acceptable to everyone?

**Comment by dwsteele** [comment]
This has been fixed at 01c2669b976.

@CluelessTechnologist if you would like to appear in the release notes with a different name, just let us know.
